polkadot/node/collation-generation/src/lib.rs

```diff
@@ -624,48 +582,23 @@ async fn construct_and_distribute_receipt(
 
 	sender
 		.send_message(CollatorProtocolMessage::DistributeCollation {
-			candidate_receipt: ccr,
+			candidate_receipt: receipt,
 			parent_head_data_hash,
 			pov,
 			parent_head_data,
 			result_sender,
 			core_index,
 		})
 		.await;
-}
 
-async fn obtain_validation_code_hash_with_assumption(
-	relay_parent: Hash,
-	para_id: ParaId,
-	assumption: OccupiedCoreAssumption,
-	sender: &mut impl overseer::CollationGenerationSenderTrait,
-) -> crate::error::Result<Option<ValidationCodeHash>> {
-	match request_validation_code_hash(relay_parent, para_id, assumption, sender)
-		.await
-		.await?
-	{
-		Ok(Some(v)) => Ok(Some(v)),
-		Ok(None) => Ok(None),
-		Err(RuntimeApiError::NotSupported { .. }) => {
-			match request_validation_code(relay_parent, para_id, assumption, sender).await.await? {
-				Ok(Some(v)) => Ok(Some(v.hash())),
-				Ok(None) => Ok(None),
-				Err(e) => {
-					// We assume that the `validation_code` API is always available, so any error
-					// is unexpected.
-					Err(e.into())
-				},
-			}
-		},
-		Err(e @ RuntimeApiError::Execution { .. }) => Err(e.into()),
-	}
+	Ok(())
 }
 
 fn erasure_root(
 	n_validators: usize,
 	persisted_validation: PersistedValidationData,
 	pov: PoV,
```

Comment:
Given the above comment, this should be receipt.core_index().unwrap_or(core_index)

Reply:
I don't understand how this helps. We are creating the receipt just above with this core index

Reply:
The main point here is that we should pick the core index that the candidate commits to. There is a bit of confusion since there are 2 core indices, one from the PreparedCollation and potentially another one in the commitments. If we are using v2 receipts, the collator code retrieves the core index from the commitments and puts it in the SubmitCollationParams and then we check it once again later here. This is likely prone to errors if the core_index is different from the one passed in the commitments. Maybe we should have a separate SubmitCollationParamsV2 that doesn't have a core_index field at all (which should be used when the candidate commits to a `core_index`), so there is no confusion.

Reply:
ok. what you're saying seems doable. However, I'm not sure it simplifies things.
We'd need to handle two message types here in collation-generation and also decide which message type to use in the collator based on the node feature and whether the core index commitment is present. I'm not sure it simplifies things.
Would be an improvement idea for when we completely switch to v2 receipts and make the commitment mandatory

Reply:
The commitments being optional indeed introduces more complexity, but for v2 receipts, the collator doesn't actually need to know the core index, just the amount of cores assigned. We don't even need a runtime API to get the next core selector, we just build a block and we look at the commitments, if there is a core committed, we just push the collation to collation generation. We can create a ticket to handle this at a later time.

Reply:
I think I remember why I coded it this way. What you're suggesting has the caveat that we're breaking the generic nature of the core selection policy of the parachain runtime, since we assume that the runtime will choose a core that is assigned. If it won't, it would have built a block and only after that see that the selected core is not assigned.